hvsock_test.go

```diff
@@ -0,0 +1,631 @@
+	l, addr := serverListen(u)
+
+	ch := make(chan struct{})
+	go func() {
+		defer close(ch)
+		c, err := l.Accept()
+		u.Must(err, "listener accept")
+		defer c.Close()
+
+		hv := c.(*HvsockConn)
+		//
+		// test CloseWrite()
+		//
+		n, err := c.Write([]byte(testStr))
+		u.Must(err, "server tx")
+		u.Assert(n == len(testStr), fmt.Sprintf("server wrote %d bytes, wanted %d", n, len(testStr)))
+
+		u.Must(hv.CloseWrite(), "server close write")
+
+		_, err = c.Write([]byte(testStr))
+		u.Is(err, windows.WSAESHUTDOWN, "server did not shutdown writes")
+
+		// safe to call multiple times
+		u.Must(hv.CloseWrite(), "server second close write")
+
+		//
+		// test CloseRead()
+		//
+		b := make([]byte, 256)
+		n, err = c.Read(b)
+		u.Must(err, "server read")
+		u.Assert(n == len(testStr), fmt.Sprintf("server read %d bytes, wanted %d", n, len(testStr)))
+		u.Assert(string(b[:n]) == testStr, fmt.Sprintf("server got %q; wanted %q", b[:n], testStr))
+
+		u.Must(hv.CloseRead(), "server close read")
+
+		ch <- struct{}{}
+
+		// signal the client to send more info
+		// if it was sent before, the read would succeed if the data was buffered prior
+		_, err = c.Read(b)
+		u.Is(err, windows.WSAESHUTDOWN, "server did not shutdown reads")
+
+		// safe to call multiple times
+		u.Must(hv.CloseRead(), "server second close read")
+
+		c.Close()
+		u.Is(hv.CloseWrite(), socket.ErrSocketClosed, "client close write")
+		u.Is(hv.CloseRead(), socket.ErrSocketClosed, "client close read")
+	}()
+
+	cl, err := Dial(context.Background(), addr)
+	u.Must(err, "could not dial")
+	defer cl.Close()
+
+	b := make([]byte, 256)
+	n, err := cl.Read(b)
+	u.Must(err, "client read")
+	u.Assert(n == len(testStr), fmt.Sprintf("client read %d bytes, wanted %d", n, len(testStr)))
+	u.Assert(string(b[:n]) == testStr, fmt.Sprintf("client got %q; wanted %q", b[:n], testStr))
+
+	n, err = cl.Read(b)
+	u.Assert(n == 0, "client did not get EOF")
+	u.Is(err, io.EOF, "client did not get EOF")
+
+	n, err = cl.Write([]byte(testStr))
+	u.Must(err, "client write")
+	u.Assert(n == len(testStr), fmt.Sprintf("client wrote %d bytes, wanted %d", n, len(testStr)))
+
+	u.Wait(ch, time.Second)
+	u.Check()
+
+	// this should succeed
+	_, err = cl.Write([]byte("test2"))
+	u.Must(err, "client write")
+}
+
+func TestHvSockCloseReadWriteDial(t *testing.T) {
+	u := newUtil(t)
+	l, addr := serverListen(u)
+
+	ch := make(chan struct{})
+	go func() {
+		defer close(ch)
+		c, err := l.Accept()
+		u.Must(err, "listener accept")
+		defer c.Close()
+
+		b := make([]byte, 256)
+		n, err := c.Read(b)
+		u.Must(err, "server read")
+		u.Assert(string(b[:n]) == testStr, fmt.Sprintf("server got %q; wanted %q", b[:n], testStr))
+
+		n, err = c.Read(b)
+		u.Assert(n == 0, "server did not get EOF")
+		u.Is(err, io.EOF, "server did not get EOF")
+
+		_, err = c.Write([]byte(testStr))
+		u.Must(err, "server tx")
+
+		ch <- struct{}{}
+
+		_, err = c.Write([]byte(testStr))
+		u.Must(err, "server tx")
+
+		c.Close()
+	}()
```

Comment:
Fatal calls in a goroutine just stop the goroutine itself and not the test. Can we have these goroutines send to an error chan inside instead and we can read on the main thread? It looks like the pattern of:

Make chan -> goroutine for listener/dial -> send/close to chan on success might also warrant a new helper for your testing utility, maybe you pass in an anonymous func that makes use of a channel and the channel in question is made in the method and returned. e.g. something like:

```go
func (h *hi) Go(run func(chan error)) chan error {
	ch := make(chan error)
	go func() {
		run(ch)
	}()
	return ch
}


errCh := u.Go(func(ch chan error){
     ch <- errors.New("oh no")
})


err <- errCh

```

Reply:
I actually toyed around with this a bunch, but in `run`, we would still have to write
```go
errCh := u.Go(func(ch chan error){
    if err := doSomething(...); err != nil {
        ch <- fmt.Errorf("oh no: %w", err)
        return
    }
    moreThings(...)
})
u.Must(<- errCh)
```

and instead `u.Must` and co. give you give you the `if err != nil { return }` for free, so you avoid a lot of boilerplate
The only downside is we assume that you always want to stop a test (`t.FailNow()`) immediately on failure instead of just calling `t.Fail()`, but I haven't see cases where we do the latter.

I have `u.Check()` so that way you can check if a go-routine failed (via `t.Failed()`), and in the cases I do need to synchronize (eg, `TestHvSockCloseReadWriteDial`), I do use a channel as a block.

Reply:
But also, I think your point is valid and this isnt the most flexible approach--I'm just hesitant to do a redesign in this PR 😅
We should have a more detailed design discussion in the future about testing primitives....

Reply:
I mean the signature could even be:

```go
func (u *util) Go(run func() error ) chan error {
	ch := make(chan error)
	go func() {
		ch <- run()
	}()
	return ch
}

func TestBlah(t *testing.T) {
	errCh := Go(func() error {
		if err := stuff(); err != nil {
			return err
		}
		return nil
	})

	u.Wait(errCh, time.Second*5)
}
```

Now you just write standard error handling for your concurrent bits, and can just pass the channel to your Wait().

Reply:
Are we just recreating [`errgroup.Group`](https://pkg.go.dev/golang.org/x/sync/errgroup#Group) here?

Reply:
> Are we just recreating [`errgroup.Group`](https://pkg.go.dev/golang.org/x/sync/errgroup#Group) here?

Kind of, but 
 1. we dont always need to wait for the goroutine to finish running; and 
 2.  the main point of the `util` is to avoid having to write:
```go
if err != nil {
    return err
}
// ...
return nil
```
or 
```go
if !errors.Is(err, target) {
    return fmt.Errorf("got error %v, wanted %v", err, target)
}
// ...
return nil
```

rarely here or in our other tests do we do anything besides ending the test when an error (or the wrong one) occurs, so it adds more overhead to constantly have to add return statements everywhere.
Especially since the main test routine will eventually just print error to stdout and then call `FailNow()`

Additionally, `errorgroup.Group` only has a `g.Wait() error` function, so to use it in a select statement (ie, to add a timeout), you need to do pass in a context. And that context's cancellation doesn't propagate to the error group, so you have to use the `ctx.Done()` or the returned context in conjunction with another timeout

Reply:
I think my main point is it feels weird to have a model of:

Kill goroutine if error ->  wait in main thread -> channel gets closed by virtue of the `defer close(ch)` so Wait wakes up -> check if it failed after Wait()

Vs.

Send nil or error to chan -> Wait in main thread for error/nil/timeout from goroutine -> t.Fatal if timeout or err != nil

Reply:
Yes, but it saves typing 😅😅
I guess your suggestion would be cleaner:
```go 
func (u *util) Go(run func() error ) chan error {
	ch := make(chan error)
	go func() {
		ch <- run()
	}()
	return ch
}

func (u testUtil) WaitErr(ch <-chan error, d time.Duration, msgs ...string) {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case err := <-ch:
		u.Must(err, msgs...)
	case <-t.C:
		u.T.Helper()
		u.T.Fatalf(msgJoin(msgs, "timed out after %v"), d)
	}
}
```

Reply:
its just irksome that `test` has a utility function to end a test early, but it cannot be used from a goroutine....